src/file-table-engine/src/engine/immutable.rs

```diff
@@ -0,0 +1,356 @@
+        self.inner.get_table(table_ref).is_some()
+    }
+
+    async fn drop_table(
+        &self,
+        _ctx: &EngineContext,
+        request: DropTableRequest,
+    ) -> TableResult<bool> {
+        self.inner
+            .drop_table(request)
+            .await
+            .map_err(BoxedError::new)
+            .context(table_error::TableOperationSnafu)
+    }
+
+    async fn close(&self) -> TableResult<()> {
+        self.inner.close().await
+    }
+}
+
+struct EngineInner {
+    /// All tables opened by the engine. Map key is formatted [TableReference].
+    ///
+    /// Writing to `tables` should also hold the `table_mutex`.
+    tables: RwLock<HashMap<String, ImmutableFileTableRef>>,
+    object_store: ObjectStore,
+
+    /// Table mutex is used to protect the operations such as creating/opening/closing
+    /// a table, to avoid things like opening the same table simultaneously.
+    table_mutex: Mutex<()>,
+}
+
+impl EngineInner {
+    async fn create_table(
+        &self,
+        _ctx: &EngineContext,
+        request: CreateTableRequest,
+    ) -> Result<TableRef> {
+        let CreateTableRequest {
+            catalog_name,
+            schema_name,
+            table_name,
+            create_if_not_exists,
+            table_options,
+            ..
+        } = request;
+        let table_ref = TableReference {
+            catalog: &catalog_name,
+            schema: &schema_name,
+            table: &table_name,
+        };
+
+        if let Some(table) = self.get_table(&table_ref) {
+            return if create_if_not_exists {
+                Ok(table)
+            } else {
+                TableExistsSnafu { table_name }.fail()
+            };
+        }
+
+        let table_schema =
+            Arc::new(Schema::try_from(request.schema).context(InvalidRawSchemaSnafu)?);
+
+        let table_id = request.id;
+        let table_dir = table_dir(&catalog_name, &schema_name, table_id);
+
+        let table_full_name = table_ref.to_string();
+
+        let _lock = self.table_mutex.lock().await;
+        // Checks again, read lock should be enough since we are guarded by the mutex.
+        if let Some(table) = self.get_table_by_full_name(&table_full_name) {
+            return if request.create_if_not_exists {
+                Ok(table)
+            } else {
+                TableExistsSnafu { table_name }.fail()
+            };
+        }
+
+        let table_meta = TableMetaBuilder::new_external_table()
+            .schema(table_schema)
+            .engine(IMMUTABLE_FILE_ENGINE)
+            .options(table_options)
+            .build()
+            .context(BuildTableMetaSnafu {
+                table_name: &table_full_name,
+            })?;
+
+        let table_info = TableInfoBuilder::new(&table_name, table_meta)
+            .ident(table_id)
+            .table_version(INIT_TABLE_VERSION)
+            .table_type(TableType::Base)
```

Comment:
As we've discussed, if external table is scoped within a session, maybe we should use `TableType::Temporary` here? It does not affect correctness tho.

Reply:
> As we've discussed, if external table is scoped within a session, maybe we should use `TableType::Temporary` here? It does not affect correctness tho.

However, In RFC, we decided to make it durable finally.🤣